### PR TITLE
[alt] add a benchmark stage

### PR DESCRIPTION
**Pull request by zjgarvey**
1. adds a stage "benchmark" which is not included by default
2. implements benchmarking for mode = cl-onnx-iree
3. adds benchmark time to generated reports
4. adds a run.py command line arg `--benchmark` for convenience. 

Sample usage:

```
python run.py -m cl-onnx-iree --benchmark -t migraphx -v --report
```

generates:

## Summary

|Stage|Count|
|--|--|
| Total | 30 |
| PASS | 12 |
| Numerics | 4 |
| results-summary | 0 |
| postprocessing | 0 |
| benchmark | 0 |
| compiled_inference | 2 |
| native_inference | 2 |
| construct_inputs | 0 |
| compilation | 7 |
| preprocessing | 0 |
| import_model | 3 |
| setup | 0 |

## Test Run Detail 
Test was run with the following arguments:
Namespace(device='local-task', backend='llvm-cpu', iree_compile_args=None, mode='cl-onnx-iree', torchtolinalg=False, stages=None, skip_stages=None, benchmark=True, load_inputs=False, groups='all', test_filter='migraphx', testsfile=None, tolerance=None, verbose=True, rundirectory='test-run', no_artifacts=False, report=True, report_file='report.md')

| Test | Exit Status | Mean Benchmark Time (ms) | Notes |
|--|--|--|--|
| migraphx_agentmodel__AgentModel | compilation | None | |
| migraphx_bert__bert-large-uncased | compilation | None | |
| migraphx_bert__bertsquad-12 | compilation | None | |
| migraphx_cadene__dpn92i1 | PASS | 527.7663788991049 | |
| migraphx_cadene__inceptionv4i16 | PASS | 25202.195480884984 | |
| migraphx_cadene__resnext101_64x4di1 | PASS | 1144.4547850056551 | |
| migraphx_cadene__resnext101_64x4di16 | PASS | 11430.92905820231 | |
| migraphx_huggingface-transformers__bert_mrpc8 | native_inference | None | |
| migraphx_mlperf__bert_large_mlperf | Numerics | 9947.680042800494 | |
| migraphx_mlperf__resnet50_v1 | PASS | 320.7237229478779 | |
| migraphx_models__whisper-tiny-decoder | compiled_inference | None | |
| migraphx_models__whisper-tiny-encoder | native_inference | None | |
| migraphx_onnx-misc__taau_low_res_downsample_d2s_for_infer_time_fp16_opset11 | import_model | None | |
| migraphx_onnx-model-zoo__gpt2-10 | compilation | None | |
| migraphx_ORT__bert_base_cased_1 | PASS | 1057.2638947051018 | |
| migraphx_ORT__bert_base_uncased_1 | compilation | None | |
| migraphx_ORT__bert_large_uncased_1 | PASS | 3485.4127474012785 | |
| migraphx_ORT__distilgpt2_1 | compiled_inference | None | |
| migraphx_ORT__onnx_models__bert_base_cased_1_fp16_gpu | Numerics | 3408.8035288092215 | |
| migraphx_ORT__onnx_models__bert_large_uncased_1_fp16_gpu | Numerics | 11613.216011907207 | |
| migraphx_ORT__onnx_models__distilgpt2_1_fp16_gpu | Numerics | 171.46043484826805 | |
| migraphx_pytorch-examples__wlang_gru | compilation | None | |
| migraphx_pytorch-examples__wlang_lstm | compilation | None | |
| migraphx_sd__unet__model | import_model | None | |
| migraphx_sdxl__unet__model | import_model | None | |
| migraphx_torchvision__densenet121i32 | PASS | 4118.973942403682 | |
| migraphx_torchvision__inceptioni1 | PASS | 718.8349567120895 | |
| migraphx_torchvision__inceptioni32 | PASS | 23589.06321739778 | |
| migraphx_torchvision__resnet50i1 | PASS | 312.4926148477243 | |
| migraphx_torchvision__resnet50i64 | PASS | 14743.020548688946 | |

### CODE REVIEW

File: alt_e2eshark/run.py [hunk 1]
```diff
@@ -228,6 +241,13 @@ def run_tests(
                 outputs_raw = config.run(compiled_artifact, inputs, func_name=func_name)
                 outputs_raw.save_to(log_dir + "output")
 
+            # benchmark inference time with compiled module
+            curr_stage = "benchmark"
+            if curr_stage in stages:
+                notify_stage()
+                # TODO: make repetitions configurable from a command line arg
+                mean_time_ms = config.benchmark(compiled_artifact, inputs, repetitions=5, func_name=func_name)
```
Comment (saienduri) [inline alt_e2eshark/run.py:249]: In interest of nightly times, maybe drop to 2 or 3 repetitions?

File: alt_e2eshark/e2e_testing/test_configs/onnxconfig.py [hunk 2]
```diff
@@ -231,3 +232,27 @@ def run(self, artifact: str, inputs: TestTensors, *, func_name=None) -> TestTens
                 raise FileNotFoundError(error_msg)
         return TestTensors.load_from(self.tensor_info_dict[test_name][0], self.tensor_info_dict[test_name][1], run_dir, "output")
 
+    def benchmark(self, artifact: str, inputs: TestTensors, repetitions: int = 5, *, func_name=None) -> float:
+        run_dir = Path(artifact).parent
+        detail_log = run_dir.joinpath("detail", "benchmark.detail.log")
+        commands_log = run_dir.joinpath("commands", "benchmark.commands.log")
+        report_json = run_dir.joinpath("detail", "benchmark.json")
```
Comment (saienduri) [inline alt_e2eshark/e2e_testing/test_configs/onnxconfig.py:239]: Can we edit cleanup flag level 3 to also include this benchmark.json file?

Comment (zjgarvey) [reply alt_e2eshark/e2e_testing/test_configs/onnxconfig.py:239]: Sure!

File: alt_e2eshark/run.py (hunk 1 above)
Comment (zjgarvey) [reply alt_e2eshark/run.py:249]: I originally had it at ten then moved to five haha. I'll drop the default to 3.

**Review by saienduri — APPROVED** [review]
Thanks!

**Pull request merged**